File: fuzzysearch/src/main/java/com/zxn/fuzzysearch/FuzzySearchBaseAdapter.java
package com.zxn.fuzzysearch;

import android.text.TextUtils;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.Nullable;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * Updated by zxn on 2020/7/7.
 */
public abstract class FuzzySearchBaseAdapter<ITEM extends IFuzzySearchItem, K extends BaseViewHolder>
        extends BaseQuickAdapter<ITEM, K> implements Filterable {

    private FuzzySearchFilter mFilter;
    private List<ITEM> mBackDataList;
    private IFuzzySearchRule mIFuzzySearchRule;

    public FuzzySearchBaseAdapter(IFuzzySearchRule rule, List<ITEM> dataList, int layoutResId) {
        super(layoutResId, dataList);
        if (rule == null) {
            mIFuzzySearchRule = new DefaultFuzzySearchRule();
        }
        mBackDataList = dataList;
    }

    @Override
    public void setNewData(@Nullable List<ITEM> data) {
        super.setNewData(data);
        mBackDataList = data;
    }

    @Override
    public Filter getFilter() {
        if (mFilter == null) {
            mFilter = new FuzzySearchFilter();
        }
        return mFilter;
    }

    private class FuzzySearchFilter extends Filter {

        /**
         * 执行过滤操作,如果搜索的关键字为空，默认所有结果
         */
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults result = new FilterResults();
            List<ITEM> filterList;
            if (TextUtils.isEmpty(constraint)) {
                filterList = mBackDataList;
            } else {
                filterList = new ArrayList<>();
                for (ITEM item : mBackDataList) {
                    if (mIFuzzySearchRule.accept(constraint, item.getSourceKey(), item.getFuzzyKey())) {
                        filterList.add(item);
                    }
                }
            }
            result.values = filterList;
            result.count = filterList.size();
            return result;
        }

        /**
         * 得到过滤结果
         */
        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            mData = (List<ITEM>) results.values;
            notifyDataSetChanged();
        }
    }

}

File: fuzzysearch/src/main/java/com/zxn/fuzzysearch/FuzzySearchBaseAdapter.kt
package com.zxn.fuzzysearch

import android.text.TextUtils
import android.widget.Filter
import android.widget.Filterable
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import java.util.*

/**
 * Updated by zxn on 2020/11/2.
 */
abstract class FuzzySearchBaseAdapter<ITEM : IFuzzySearchItem, K : BaseViewHolder>(
    rule: IFuzzySearchRule?,
    dataList: MutableList<ITEM>?,
    layoutResId: Int
) : BaseQuickAdapter<ITEM, K>(layoutResId, dataList), Filterable {
    private var mFilter: FuzzySearchFilter? = null
    private var mBackDataList: MutableList<ITEM>?
    private var mIFuzzySearchRule: IFuzzySearchRule? = null

    init {
        if (rule == null) {
            mIFuzzySearchRule = DefaultFuzzySearchRule()
        }
        mBackDataList = dataList
    }


    override fun setList(list: Collection<ITEM>?) {
        super.setList(list)
        mBackDataList = data
    }

    override fun getFilter(): Filter? {
        if (mFilter == null) {
            mFilter = FuzzySearchFilter()
        }
        return mFilter
    }

    private inner class FuzzySearchFilter : Filter() {
        /**
         * 执行过滤操作,如果搜索的关键字为空，默认所有结果
         */
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val result = FilterResults()
            val filterList: MutableList<ITEM>?
            if (TextUtils.isEmpty(constraint)) {
                filterList = mBackDataList
            } else {
                filterList = ArrayList()
                for (item in mBackDataList!!) {
                    if (mIFuzzySearchRule!!.accept(constraint, item!!.sourceKey, item.fuzzyKey)) {
                        filterList.add(item)
                    }
                }
            }
            result.values = filterList
            result.count = filterList!!.size
            return result
        }

        /**
         * 得到过滤结果
         */
        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            //setNewData(results.values as MutableList<ITEM>)
            setList(results.values as MutableList<ITEM>)
            notifyDataSetChanged()
        }
    }



//    override fun setNewData(data: MutableList<ITEM>?) {
//        super.setNewData(data)
//        mBackDataList = data
//    }
}